compiler/rustc_hir_typeck/src/op.rs

```diff
@@ -776,7 +775,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             (None, Some(trait_did)) => {
                 let (obligation, _) =
                     self.obligation_for_method(cause, trait_did, lhs_ty, Some(input_types));
-                Err(rustc_trait_selection::traits::fully_solve_obligation(self, obligation))
+                // FIXME: This should potentially just add the obligation to the `FnCtxt`
+                let ocx = ObligationCtxt::new(&self.infcx);
+                ocx.register_obligation(obligation);
+                Err(ocx.select_all_or_error())
```

Comment:
We probably use these errors eagerly for better diagnostics. It's likely possible to re-create this `obligation_for_method` in the diagnostics code and re-process it though, or move the diagnostics code into the `ObligationCauseCode::BinOp` handling branch of `note_cause_code` or whatever.